readers/hdfs/src/main/java/com/criteo/hadoop/garmadon/hdfs/writer/PartitionedWriter.java

```diff
@@ -0,0 +1,237 @@
+
+    private final Function<LocalDateTime, ExpiringConsumer<MessageKind>> writerBuilder;
+    private final OffsetComputer offsetComputer;
+    private final Map<Integer, Map<LocalDateTime, ExpiringConsumer<MessageKind>>> perPartitionDayWriters;
+    private final HashMap<Integer, Long> perPartitionStartOffset;
+    private final FileNamer fileNamer;
+
+    /**
+     *
+     * @param writerBuilder     Builds an expiring writer based on a path.
+     */
+    public PartitionedWriter(Function<LocalDateTime, ExpiringConsumer<MessageKind>> writerBuilder, OffsetComputer offsetComputer,
+                      FileNamer fileNamer) {
+        this.writerBuilder = writerBuilder;
+        this.offsetComputer = offsetComputer;
+        this.perPartitionDayWriters = new HashMap<>();
+        this.perPartitionStartOffset = new HashMap<>();
+        this.fileNamer = fileNamer;
+    }
+
+    /**
+     * Stops processing events for a given partition.
+     *
+     * @param partitionId   The partition for which to stop processing events.
+     */
+    public void dropPartition(int partitionId) {
+        synchronized (perPartitionDayWriters) {
+            perPartitionDayWriters.keySet().remove(partitionId);
+            perPartitionStartOffset.keySet().remove(partitionId);
+        }
+    }
+
+    /**
+     * Write a message to a dedicated file
+     * File path: day.partitionId.firstMessageOffset, eg. 1987-08-13.11.101
+     *
+     * @param when          Message time, used to route to the correct file
+     * @param offset        Message offset, used to route to the correct file
+     * @param msg           Message to be written
+     */
+    public void write(Instant when, Offset offset, MessageKind msg) throws IOException {
+        LocalDateTime dayStartTime = LocalDateTime.ofInstant(when.truncatedTo(ChronoUnit.DAYS), UTC_ZONE);
+        int partitionId = offset.getPartition();
+
+        synchronized (perPartitionDayWriters) {
+            long startingOffset = getStartingOffset(partitionId);
+
+            if (offset.getOffset() <= startingOffset)
+                return;
+
+            String path = fileNamer.buildPath(dayStartTime, offset);
+
+            // /!\ This line must not be switched with the offset computation as this would create empty files otherwise
+            ExpiringConsumer<MessageKind> consumer = getWriter(dayStartTime, partitionId, path);
+
+            consumer.write(msg, offset);
+        }
+    }
+
+    /**
+     * Close all open consumers
+     */
+    @Override
+    public void close() {
+        possiblyCloseConsumers(ignored -> true);
+    }
+
+    /**
+     * Get the starting offset for a given partition.
+     *
+     * @param partitionId
+     * @return              The lowest offset for a given partition
+     * @throws IOException  If the offset computation failed
+     */
+    public long getStartingOffset(int partitionId) throws IOException {
```

Comment:
this public method is possibly called by multiple threads, and therefore not thread safe!